crates/shim/src/mount.rs

```diff
@@ -223,6 +231,286 @@ lazy_static! {
+    fn from(err: nix::errno::Errno) -> Self {
+        match err {
+            nix::errno::Errno::UnknownErrno => MountExitCode::NixUnknownErr,
+            _ => MountExitCode::NixOtherErr(err as i32),
+        }
+    }
+}
+
+impl From<MountExitCode> for nix::errno::Errno {
+    fn from(code: MountExitCode) -> Self {
+        match code {
+            MountExitCode::NixOtherErr(errno) => nix::errno::Errno::from_i32(errno),
+            _ => nix::errno::Errno::UnknownErrno,
+        }
+    }
+}
+
+impl From<MountExitCode> for Result<()> {
+    fn from(code: MountExitCode) -> Self {
+        match code {
+            MountExitCode::NixUnknownErr => Err(mount_process_error!(
+                "mount process exit unexpectly, exit code: {}",
+                nix::errno::Errno::from(code)
+            )),
+            MountExitCode::ChdirErr => Err(mount_process_error!(
+                "mount process exit unexpectly: chdir failed"
+            )),
+            MountExitCode::Success => Ok(()),
+            MountExitCode::NixOtherErr(errno) => Err(mount_process_error!(
+                "mount process exit unexpectly, exit code: {}",
+                nix::errno::Errno::from_i32(errno)
+            )),
+        }
+    }
+}
+
+#[cfg(not(feature = "async"))]
+#[cfg(target_os = "linux")]
+pub fn mount_rootfs(
+    fs_type: Option<&str>,
+    source: Option<&str>,
+    options: &[String],
+    target: impl AsRef<Path>,
+) -> Result<()> {
+    //TODO add helper to mount fuse
+    let max_size = page_size::get();
+    // avoid hitting one page limit of mount argument buffer
+    //
+    // NOTE: 512 id a buffer during pagesize check.
+    let (chdir, options) =
+        if fs_type.unwrap_or("") == "overlay" && options_size(options) >= max_size - 512 {
+            LowerdirCompactor::new(options).compact()
+        } else {
+            (None, options.to_vec())
+        };
+
+    let mut flags: MsFlags = MsFlags::from_bits(0).unwrap();
+    let mut data = Vec::new();
+    options.iter().for_each(|x| {
+        if let Some(f) = MOUNT_FLAGS.get(x.as_str()) {
+            if f.clear {
+                flags.bitand_assign(f.flags.not());
+            } else {
+                flags.bitor_assign(f.flags)
+            }
+        } else {
+            data.push(x.as_str())
+        }
+    });
+
+    let opt = data.join(",");
+    if opt.len() > max_size {
+        return Err(mount_process_error!("mount option is too long"));
+    }
+
+    let data = if !data.is_empty() {
+        Some(opt.as_str())
+    } else {
+        None
+    };
+
+    let s = monitor_subscribe(Topic::All)?;
+    match unsafe { fork() } {
+        Ok(ForkResult::Parent { child, .. }) => {
+            let code: MountExitCode = wait_pid(i32::from(child), s).into();
+            code.into()
+        }
+        Ok(ForkResult::Child) => {
+            if let Some(workdir) = chdir {
+                env::set_current_dir(Path::new(&workdir)).unwrap_or_else(|_| {
```

Comment:
I think it's no need to convert err code, you can print the detailed error log msg in child process and exit with none-zero code. We could check the log file for details if any error.

Reply:
add detailed error log msg in child process.